File: backend/src/services/ai-ops/rag/migration/hybridSearchMigration.ts
/**
 * HybridSearchMigration - 混合检索迁移工具
 * 
 * 用于迁移现有知识条目，增强元数据并重建关键词索引。
 * 
 * Requirements: 5.6
 * - 迁移所有现有条目
 * - 支持增量迁移
 * - 验证迁移结果
 */

import { logger } from '../../../../utils/logger';
import { KnowledgeBase, KnowledgeEntry } from '../knowledgeBase';
import { MetadataEnhancer } from '../metadataEnhancer';
import { KeywordIndexManager } from '../keywordIndexManager';
import {
  MigrationResult,
  MigrationProgress,
  VerificationResult,
  EnhancedMetadata,
} from '../types/hybridSearch';

/**
 * 迁移配置
 */
export interface MigrationConfig {
  /** 批量处理大小 */
  batchSize: number;
  /** 是否跳过已增强的条目 */
  skipEnhanced: boolean;
  /** 是否强制重新增强 */
  forceReEnhance: boolean;
  /** 进度回调间隔 */
  progressInterval: number;
}

/**
 * 默认迁移配置
 */
export const DEFAULT_MIGRATION_CONFIG: MigrationConfig = {
  batchSize: 10,
  skipEnhanced: true,
  forceReEnhance: false,
  progressInterval: 5,
};

/**
 * HybridSearchMigration 迁移工具类
 */
export class HybridSearchMigration {
  private knowledgeBase: KnowledgeBase;
  private metadataEnhancer: MetadataEnhancer;
  private keywordIndexManager: KeywordIndexManager;
  private config: MigrationConfig;

  constructor(
    knowledgeBase: KnowledgeBase,
    metadataEnhancer?: MetadataEnhancer,
    keywordIndexManager?: KeywordIndexManager,
    config?: Partial<MigrationConfig>
  ) {
    this.knowledgeBase = knowledgeBase;
    this.metadataEnhancer = metadataEnhancer || knowledgeBase.getMetadataEnhancer()!;
    this.keywordIndexManager = keywordIndexManager || knowledgeBase.getKeywordIndexManager()!;
    this.config = { ...DEFAULT_MIGRATION_CONFIG, ...config };

    if (!this.metadataEnhancer) {
      throw new Error('MetadataEnhancer not available');
    }
    if (!this.keywordIndexManager) {
      throw new Error('KeywordIndexManager not available');
    }
  }

  /**
   * 迁移所有现有条目
   * Requirements: 5.6
   * 
   * @param onProgress 进度回调
   * @returns 迁移结果
   */
  async migrateAll(
    onProgress?: (progress: MigrationProgress) => void
  ): Promise<MigrationResult> {
    const startTime = Date.now();
    const entries = this.knowledgeBase.getAllEntries();
    
    logger.info('Starting hybrid search migration', { totalEntries: entries.length });

    const result: MigrationResult = {
      total: entries.length,
      success: 0,
      failed: 0,
      skipped: 0,
      failedIds: [],
      duration: 0,
    };

    // 过滤需要迁移的条目
    const entriesToMigrate = this.config.skipEnhanced && !this.config.forceReEnhance
      ? entries.filter(e => !e.metadata.enhancedAt)
      : entries;

    result.skipped = entries.length - entriesToMigrate.length;

    // 分批处理
    for (let i = 0; i < entriesToMigrate.length; i += this.config.batchSize) {
      const batch = entriesToMigrate.slice(i, i + this.config.batchSize);
      
      // 处理当前批次
      const batchResults = await Promise.allSettled(
        batch.map(entry => this.migrateEntry(entry.id))
      );

      // 统计结果
      for (let j = 0; j < batchResults.length; j++) {
        const batchResult = batchResults[j];
        if (batchResult.status === 'fulfilled') {
          result.success++;
        } else {
          result.failed++;
          result.failedIds.push(batch[j].id);
          logger.error('Migration failed for entry', {
            entryId: batch[j].id,
            error: batchResult.reason,
          });
        }
      }

      // 报告进度
      if (onProgress && (i + this.config.batchSize) % (this.config.progressInterval * this.config.batchSize) === 0) {
        const current = Math.min(i + this.config.batchSize, entriesToMigrate.length);
        const elapsed = Date.now() - startTime;
        const avgTimePerEntry = elapsed / current;
        const remaining = entriesToMigrate.length - current;
        
        onProgress({
          current: current + result.skipped,
          total: entries.length,
          success: result.success,
          failed: result.failed,
          percentage: Math.round(((current + result.skipped) / entries.length) * 100),
          estimatedRemaining: Math.round(avgTimePerEntry * remaining),
        });
      }
    }

    // 持久化关键词索引
    await this.keywordIndexManager.persist();

    result.duration = Date.now() - startTime;

    logger.info('Hybrid search migration completed', {
      total: result.total,
      success: result.success,
      failed: result.failed,
      skipped: result.skipped,
      duration: result.duration,
    });

    return result;
  }

  /**
   * 迁移单个条目
   * Requirements: 5.6
   * 
   * @param entryId 条目 ID
   */
  async migrateEntry(entryId: string): Promise<void> {
    const entry = await this.knowledgeBase.get(entryId);
    if (!entry) {
      throw new Error(`Entry ${entryId} not found`);
    }

    // 增强元数据
    const enhanced = await this.metadataEnhancer.enhance(entry);

    // 更新条目
    await this.knowledgeBase.update(entryId, {
      metadata: {
        ...entry.metadata,
        autoKeywords: enhanced.autoKeywords,
        questionExamples: enhanced.questionExamples,
        autoSynonyms: enhanced.autoSynonyms,
        searchableText: enhanced.searchableText,
        enhancedAt: enhanced.enhancedAt,
        enhancementSource: enhanced.enhancementSource,
      },
    });

    // 更新关键词索引
    this.keywordIndexManager.updateEntry(entryId, {
      title: entry.title,
      content: entry.content,
      tags: entry.metadata.tags || [],
      autoKeywords: enhanced.autoKeywords,
      questionExamples: enhanced.questionExamples,
    });

    logger.debug('Migrated entry', { entryId });
  }

  /**
   * 验证迁移结果
   * Requirements: 5.6
   * 
   * @returns 验证结果
   */
  async verify(): Promise<VerificationResult> {
    const entries = this.knowledgeBase.getAllEntries();
    const issues: string[] = [];

    let enhancedCount = 0;
    let indexedCount = 0;

    for (const entry of entries) {
      // 检查元数据增强
      if (entry.metadata.enhancedAt) {
        enhancedCount++;
        
        // 验证增强数据完整性
        if (!entry.metadata.autoKeywords || entry.metadata.autoKeywords.length === 0) {
          issues.push(`Entry ${entry.id} has no autoKeywords`);
        }
      } else {
        issues.push(`Entry ${entry.id} not enhanced`);
      }

      // 检查关键词索引
      if (this.keywordIndexManager.hasEntry(entry.id)) {
        indexedCount++;
      } else {
        issues.push(`Entry ${entry.id} not in keyword index`);
      }
    }

    const passed = issues.length === 0;

    logger.info('Migration verification completed', {
      passed,
      totalEntries: entries.length,
      enhancedEntries: enhancedCount,
      indexedEntries: indexedCount,
      issuesCount: issues.length,
    });

    return {
      passed,
      totalEntries: entries.length,
      enhancedEntries: enhancedCount,
      indexedEntries: indexedCount,
      issues,
    };
  }

  /**
   * 检测未增强的条目
   * 
   * @returns 未增强的条目 ID 列表
   */
  getUnenhancedEntries(): string[] {
    const entries = this.knowledgeBase.getAllEntries();
    return entries
      .filter(e => !e.metadata.enhancedAt)
      .map(e => e.id);
  }

  /**
   * 获取迁移统计
   */
  getMigrationStats(): {
    total: number;
    enhanced: number;
    indexed: number;
    pending: number;
  } {
    const entries = this.knowledgeBase.getAllEntries();
    const enhanced = entries.filter(e => e.metadata.enhancedAt).length;
    const indexed = entries.filter(e => this.keywordIndexManager.hasEntry(e.id)).length;

    return {
      total: entries.length,
      enhanced,
      indexed,
      pending: entries.length - enhanced,
    };
  }
}

File: opsevo-python/src/opsevo/services/rag/hybrid_search_migration.py
"""HybridSearchMigration — migrate existing knowledge entries with metadata enhancement.

Requirements: 10.4
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any, Callable

from opsevo.services.rag.keyword_index import KeywordIndexManager
from opsevo.services.rag.metadata_enhancer import MetadataEnhancer
from opsevo.data.datastore import DataStore
from opsevo.utils.logger import get_logger

logger = get_logger(__name__)


@dataclass
class MigrationConfig:
    batch_size: int = 10
    skip_enhanced: bool = True
    force_re_enhance: bool = False
    progress_interval: int = 5


@dataclass
class MigrationResult:
    total: int = 0
    success: int = 0
    failed: int = 0
    skipped: int = 0
    failed_ids: list[str] = field(default_factory=list)
    duration: int = 0


@dataclass
class MigrationProgress:
    current: int = 0
    total: int = 0
    success: int = 0
    failed: int = 0
    percentage: int = 0
    estimated_remaining: int = 0


@dataclass
class VerificationResult:
    passed: bool = True
    total_entries: int = 0
    enhanced_entries: int = 0
    indexed_entries: int = 0
    issues: list[str] = field(default_factory=list)


class HybridSearchMigration:
    def __init__(
        self,
        datastore: DataStore,
        enhancer: MetadataEnhancer,
        keyword_index: KeywordIndexManager,
        config: MigrationConfig | None = None,
    ):
        self._ds = datastore
        self._enhancer = enhancer
        self._keyword_index = keyword_index
        self._config = config or MigrationConfig()

    async def migrate_all(
        self, on_progress: Callable[[MigrationProgress], None] | None = None
    ) -> MigrationResult:
        start = int(time.time() * 1000)
        entries = await self._ds.query("SELECT id, title, content, metadata FROM knowledge_embeddings")
        result = MigrationResult(total=len(entries))

        to_migrate = entries
        if self._config.skip_enhanced and not self._config.force_re_enhance:
            to_migrate = [e for e in entries if not (e.get("metadata") or {}).get("enhancedAt")]
            result.skipped = len(entries) - len(to_migrate)

        for i in range(0, len(to_migrate), self._config.batch_size):
            batch = to_migrate[i : i + self._config.batch_size]
            for entry in batch:
                try:
                    await self.migrate_entry(entry)
                    result.success += 1
                except Exception:
                    result.failed += 1
                    result.failed_ids.append(entry.get("id", ""))
                    logger.error("migration_entry_failed", entry_id=entry.get("id"))

            if on_progress and (i + self._config.batch_size) % (self._config.progress_interval * self._config.batch_size) == 0:
                current = min(i + self._config.batch_size, len(to_migrate)) + result.skipped
                elapsed = int(time.time() * 1000) - start
                avg = elapsed / max(current, 1)
                remaining = len(to_migrate) - min(i + self._config.batch_size, len(to_migrate))
                on_progress(MigrationProgress(
                    current=current, total=len(entries),
                    success=result.success, failed=result.failed,
                    percentage=round(current / max(len(entries), 1) * 100),
                    estimated_remaining=round(avg * remaining),
                ))

        await self._keyword_index.persist()
        result.duration = int(time.time() * 1000) - start
        logger.info("migration_completed", **{
            "total": result.total, "success": result.success,
            "failed": result.failed, "skipped": result.skipped,
            "duration": result.duration,
        })
        return result

    async def migrate_entry(self, entry: dict[str, Any]) -> None:
        enhanced = await self._enhancer.enhance(entry)
        meta = entry.get("metadata") or {}
        meta.update({
            "autoKeywords": enhanced.auto_keywords,
            "questionExamples": enhanced.question_examples,
            "autoSynonyms": enhanced.auto_synonyms,
            "searchableText": enhanced.searchable_text,
            "enhancedAt": enhanced.enhanced_at,
            "enhancementSource": enhanced.enhancement_source,
        })
        import json
        await self._ds.execute(
            "UPDATE knowledge_embeddings SET metadata = $1 WHERE id = $2",
            (json.dumps(meta, ensure_ascii=False), entry["id"]),
        )
        self._keyword_index.update_entry(entry["id"], {
            "title": entry.get("title", ""),
            "content": entry.get("content", ""),
            "tags": (entry.get("metadata") or {}).get("tags", []),
            "autoKeywords": enhanced.auto_keywords,
            "questionExamples": enhanced.question_examples,
        })

    async def verify(self) -> VerificationResult:
        entries = await self._ds.query("SELECT id, metadata FROM knowledge_embeddings")
        issues: list[str] = []
        enhanced_count = 0
        indexed_count = 0
        for e in entries:
            meta = e.get("metadata") or {}
            if meta.get("enhancedAt"):
                enhanced_count += 1
                if not meta.get("autoKeywords"):
                    issues.append(f"Entry {e['id']} has no autoKeywords")
            else:
                issues.append(f"Entry {e['id']} not enhanced")
            if self._keyword_index.has_entry(e["id"]):
                indexed_count += 1
            else:
                issues.append(f"Entry {e['id']} not in keyword index")
        return VerificationResult(
            passed=len(issues) == 0,
            total_entries=len(entries),
            enhanced_entries=enhanced_count,
            indexed_entries=indexed_count,
            issues=issues,
        )

    async def get_unenhanced_entries(self) -> list[str]:
        entries = await self._ds.query("SELECT id, metadata FROM knowledge_embeddings")
        return [e["id"] for e in entries if not (e.get("metadata") or {}).get("enhancedAt")]

    async def get_migration_stats(self) -> dict[str, int]:
        entries = await self._ds.query("SELECT id, metadata FROM knowledge_embeddings")
        enhanced = sum(1 for e in entries if (e.get("metadata") or {}).get("enhancedAt"))
        indexed = sum(1 for e in entries if self._keyword_index.has_entry(e["id"]))
        return {"total": len(entries), "enhanced": enhanced, "indexed": indexed, "pending": len(entries) - enhanced}
